### PR TITLE
[BL-303] Gohan healthcheck endpoint

### PR DESCRIPTION
**Pull request by adrianstepinski-cl**

### CODE REVIEW

File: healthcheck/healthcheck.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+type Healthcheck struct {
+	DataStore db.DB
+	Etcd      sync.Sync
+}
+
+func (healthcheck Healthcheck) DoHealthcheck() bool {
+	err1 := db.WithinTx(healthcheck.DataStore, func(transaction transaction.Transaction) error {
+		return nil
+	})
+	_, err2 := healthcheck.Etcd.Fetch("/gohan/cluster/lock/state_watch")
```
Comment (p-kozlowski) [inline healthcheck/healthcheck.go]: this key may not be the right choice: if the lock is lost, the key will disappear from ETCD and all healthchecks will return 503.
I suggest just `/gohan` instead

File: server/server_test.go [hunk 2]
```diff
@@ -1488,6 +1493,83 @@ var _ = Describe("Server package test", func() {
 			testURL("GET", versionURL, "", nil, http.StatusUnauthorized)
 		})
 	})
+
+	Describe("Healthcheck route is working", func() {
+		BeforeEach(func() {
+			err := server.GetSync().Update("/gohan/cluster/lock/state_watch", "{}")
```
[the remark this answers is not in the document]
Comment (adrianstepinski-cl) [reply server/server_test.go]: actually, it is no longer needed

File: server/server_test.go [hunk 3]
```diff
@@ -1488,6 +1493,83 @@ var _ = Describe("Server package test", func() {
...
+			if err != nil {
+				panic("healthcheck test setup failed")
+			}
+		})
+
+		AfterEach(func() {
+			err := server.GetSync().Delete("/gohan/cluster/lock/state_watch", false)
+			if err != nil {
+				panic("healthcheck test teardown failed")
```
[the remark this answers is not in the document]
Comment (adrianstepinski-cl) [reply server/server_test.go]: actually, it is no longer needed

File: server/server_test.go [hunk 4]
```diff
@@ -1488,6 +1493,83 @@ var _ = Describe("Server package test", func() {
...
+
+			mockDB.EXPECT().Options().Return(opts)
+			mockDB.EXPECT().BeginTx().Return(nil, errors.New("db closed"))
+		})
+
+		AfterEach(func() {
+			server.Healthcheck.DataStore = database
+		})
+
+		It("should return 503 if db works for members", func() {
```
[the remark this answers is not in the document]
Comment (adrianstepinski-cl) [reply server/server_test.go]: changed descriptions

File: server/server_test.go [hunk 5]
```diff
@@ -1488,6 +1493,83 @@ var _ = Describe("Server package test", func() {
...
+			server.Healthcheck.Etcd = mockSync
+
+			mockSync.EXPECT().Fetch("/gohan/cluster/lock/state_watch").Return(nil, errors.New("etcd is down"))
+		})
+
+		AfterEach(func() {
+			server.Healthcheck.Etcd = etcd
+		})
+
+		It("should return 503 if db works for members", func() {
```
[the remark this answers is not in the document]
Comment (adrianstepinski-cl) [reply server/server_test.go]: changed descriptions

File: healthcheck/healthcheck.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+func (healthcheck *Healthcheck) SetDb(db db.DB) {
+	healthcheck.DataStore = db
+}
+
+func (healthcheck *Healthcheck) SetSync(etcd sync.Sync) {
+	healthcheck.Etcd = etcd
+}
+
+func (healthcheck *Healthcheck) SetupHealthcheck(db db.DB, etcd sync.Sync) {
```
[the remark this answers is not in the document]
Comment (adrianstepinski-cl) [reply healthcheck/healthcheck.go]: changed

File: healthcheck/healthcheck.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+func (healthcheck Healthcheck) DoHealthcheck() bool {
+	err1 := db.WithinTx(healthcheck.DataStore, func(transaction transaction.Transaction) error {
+		return nil
+	})
+	_, err2 := healthcheck.Etcd.Fetch("/gohan/cluster/lock/state_watch")
+	return err1 == nil && err2 == nil
+}
+
+func (healthcheck *Healthcheck) SetDb(db db.DB) {
```
[the remark this answers is not in the document]
Comment (adrianstepinski-cl) [reply healthcheck/healthcheck.go]: removed

File: healthcheck/healthcheck.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+	DoHealthcheck() bool
+}
+
+type Healthcheck struct {
+	DataStore db.DB
+	Etcd      sync.Sync
+}
+
+func (healthcheck Healthcheck) DoHealthcheck() bool {
+	err1 := db.WithinTx(healthcheck.DataStore, func(transaction transaction.Transaction) error {
```
[the remark this answers is not in the document]
Comment (adrianstepinski-cl) [reply healthcheck/healthcheck.go]: changed

File: server/server.go [hunk 9]
```diff
@@ -77,6 +79,8 @@ type Server struct {
 func (server *Server) mapRoutes() {
 	config := util.GetConfig()
 	schemaManager := schema.GetManager()
+	server.Healthcheck.SetupHealthcheck(server.db, server.sync)
```
[the remark this answers is not in the document]
Comment (adrianstepinski-cl) [reply server/server.go]: moved

File: healthcheck/healthcheck.go [hunk 10]
```diff
@@ -0,0 +1,37 @@
+package healthcheck
+
+import (
+	"github.com/cloudwan/gohan/db"
+	"github.com/cloudwan/gohan/db/transaction"
+	"github.com/cloudwan/gohan/sync"
+)
+
+type Healthchecker interface {
+	DoHealthcheck() bool
```
[the remark this answers is not in the document]
Comment (adrianstepinski-cl) [reply healthcheck/healthcheck.go]: changed

File: healthcheck/healthcheck.go [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+	DataStore db.DB
+	Etcd      sync.Sync
+}
+
+func (healthcheck Healthcheck) DoHealthcheck() bool {
+	err1 := db.WithinTx(healthcheck.DataStore, func(transaction transaction.Transaction) error {
+		return nil
+	})
+	_, err2 := healthcheck.Etcd.Fetch("/gohan/cluster/lock/state_watch")
+	return err1 == nil && err2 == nil
```
[the remark this answers is not in the document]
Comment (adrianstepinski-cl) [reply healthcheck/healthcheck.go]: changed

File: healthcheck/healthcheck.go [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+type Healthcheck struct {
+	DataStore db.DB
+	Etcd      sync.Sync
+}
+
+func (healthcheck Healthcheck) DoHealthcheck() bool {
+	err1 := db.WithinTx(healthcheck.DataStore, func(transaction transaction.Transaction) error {
+		return nil
```
[the remark this answers is not in the document]
Comment (adrianstepinski-cl) [reply healthcheck/healthcheck.go]: done in routing function

File: healthcheck/healthcheck.go (hunk 1 above)
Comment (adrianstepinski-cl) [reply healthcheck/healthcheck.go]: using `/gohan` now

File: docs/configuration.md [hunk 13]
```diff
@@ -501,6 +501,12 @@ WARNING: The value of watched etcd keys must be a JSON dictionary.
         timing: "*/5 * * * * *"
 ```
 
+- healthcheck
+  
+  Gohan allows to determine whether everything is up and running or not.
```
Comment (zimnx) [inline docs/configuration.md]: Elaborate on what is 'everything'

File: server/api.go [hunk 14]
```diff
@@ -634,3 +635,27 @@ func mapChildNamespaceRoute(route martini.Router, namespace *schema.Namespace) {
...
+
+	log.Debug("Registering the healthcheck handler on %s", url)
+
+	get := func(w http.ResponseWriter, params martini.Params, auth schema.Authorization) {
+		defer resources.MeasureRequestTime(time.Now(), "get", healthcheckSchema.ID)
+
+		if err := healthcheck.IsHealthy(); err == nil {
+			http.Error(w, "", http.StatusOK)
+		} else {
+			log.Error("healthcheck error:" + err.Error())
```
Comment (p-kozlowski) [inline server/api.go]: please use `Errorf` for log messages with arguments (if the underlying implementation is correct, it will be more efficient)

File: tools/dev_setup.sh [hunk 15]
```diff
@@ -3,6 +3,6 @@
 go get -u github.com/kardianos/govendor
 go get -u golang.org/x/lint/golint
 go get -u github.com/gordonklaus/ineffassign
-go get -u github.com/jteeuwen/go-bindata/go-bindata
+go get -u github.com/kevinburke/go-bindata/go-bindata
```
Comment (p-kozlowski) [inline tools/dev_setup.sh]: this should be part of a different PR

File: server/api.go [hunk 16]
```diff
@@ -634,3 +635,27 @@ func mapChildNamespaceRoute(route martini.Router, namespace *schema.Namespace) {
...
+	}
+	url := healthcheckSchema.GetPluralURL()
+
+	log.Debug("Registering the healthcheck handler on %s", url)
+
+	get := func(w http.ResponseWriter, params martini.Params, auth schema.Authorization) {
+		defer resources.MeasureRequestTime(time.Now(), "get", healthcheckSchema.ID)
+
+		if err := healthcheck.IsHealthy(); err == nil {
+			http.Error(w, "", http.StatusOK)
```
Comment (zimnx) [inline server/api.go]: w.WriteHeader(http.StatusOK)

File: server/server.go [hunk 17]
```diff
@@ -400,6 +403,7 @@ func NewServer(configFile string) (*Server, error) {
 			SkipLogging: true,
 		}))
 	}
+	server.Healthcheck = *healthcheck.NewHealthcheck(server.db, server.sync)
```
Comment (zimnx) [inline server/server.go]: why not just keep it as a pointer in server struct?

File: server/server_test.go [hunk 18]
```diff
@@ -1488,6 +1493,71 @@ var _ = Describe("Server package test", func() {
...
 	})
+
+	Describe("Healthcheck", func() {
+		Context("DB and sync working", func() {
+			It("should return 200 for members", func() {
+				testURL("GET", healthcheckURL, memberTokenID, nil, http.StatusOK)
+			})
+
+			It("should return 401 for non-members", func() {
+				testURL("GET", healthcheckURL, "", nil, http.StatusUnauthorized)
```
Comment (zimnx) [inline server/server_test.go]: should authorization be required for health check?

Comment (adrianstepinski-cl) [reply server/server_test.go]: solved

File: healthcheck/healthcheck.go [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/cloudwan/gohan/db"
+	"github.com/cloudwan/gohan/db/transaction"
+	"github.com/cloudwan/gohan/sync"
+)
+
+type Healthchecker interface {
+	DoHealthcheck() bool
+}
+
+type Healthcheck struct {
```
[the remark this answers is not in the document]
Comment (adrianstepinski-cl) [reply healthcheck/healthcheck.go]: It makes this type unusable in server struct, so I think the answer is "no".

File: server/server.go (hunk 17 above)
Comment (adrianstepinski-cl) [reply server/server.go]: changed

File: server/api.go (hunk 16 above)
Comment (adrianstepinski-cl) [reply server/api.go]: changed

File: docs/configuration.md (hunk 13 above)
Comment (adrianstepinski-cl) [reply docs/configuration.md]: changed

File: tools/dev_setup.sh (hunk 15 above)
Comment (adrianstepinski-cl) [reply tools/dev_setup.sh]: done

File: etc/schema/gohan.json [hunk 20]
```diff
@@ -9,6 +9,15 @@
...
             "scope": ["admin"]
+        },
+        {
+            "action": "healthcheck",
+            "effect": "allow",
+            "id": "healthcheck",
+            "resource": {
+                "path": "/gohan/v0.1/healthcheck"
+            },
+            "principal": "Nobody"
```
[the remark this answers is not in the document]
Comment (adrianstepinski-cl) [reply etc/schema/gohan.json]: done

File: server/api.go (hunk 14 above)
Comment (adrianstepinski-cl) [reply server/api.go]: changed format

File: server/server.go [hunk 21]
```diff
@@ -19,6 +19,7 @@ import (
 	"context"
 	"crypto/tls"
 	"fmt"
+	"github.com/cloudwan/gohan/healthcheck"
```
Comment (jkmar) [inline server/server.go]: go imports

File: server/server_test.go [hunk 22]
```diff
@@ -1488,6 +1493,71 @@ var _ = Describe("Server package test", func() {
 			testURL("GET", versionURL, "", nil, http.StatusUnauthorized)
 		})
 	})
+
+	Describe("Healthcheck", func() {
+		Context("DB and sync working", func() {
```
Comment (jkmar) [inline server/server_test.go]: I'd rather used `DescribeTable` for that

File: server/server_test.go [hunk 23]
```diff
@@ -1488,6 +1493,71 @@ var _ = Describe("Server package test", func() {
...
+			It("should return 503 for members", func() {
+				testURL("GET", healthcheckURL, memberTokenID, nil, http.StatusServiceUnavailable)
+			})
+
+			It("should return 401 for non-members", func() {
+				testURL("GET", healthcheckURL, "", nil, http.StatusUnauthorized)
+			})
+		})
+	})
+
```
Comment (jkmar) [inline server/server_test.go]: dd

File: healthcheck/healthcheck.go [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+	DataStore db.DB
+	Etcd      sync.Sync
+}
+
+func (healthcheck Healthcheck) IsHealthy() error {
+	if dbErr := db.WithinTx(healthcheck.DataStore, func(transaction transaction.Transaction) error { return nil }); dbErr != nil {
+		return dbErr
+	}
+	if _, syncErr := healthcheck.Etcd.Fetch("/gohan"); syncErr != nil {
```
[the remark this answers is not in the document]
Comment (adrianstepinski-cl) [reply healthcheck/healthcheck.go]: done

File: server/api.go [hunk 25]
```diff
@@ -634,3 +635,27 @@ func mapChildNamespaceRoute(route martini.Router, namespace *schema.Namespace) {
 		},
 	)
 }
+
+func mapHealthcheckRoute(route martini.Router, manager *schema.Manager, healthcheck *healthcheck.Healthcheck) {
+	healthcheckSchema, hasSchema := manager.Schema("healthcheck")
```
[the remark this answers is not in the document]
Comment (adrianstepinski-cl) [reply server/api.go]: done

File: server/server.go (hunk 21 above)
Comment (adrianstepinski-cl) [reply server/server.go]: done

File: server/server_test.go (hunk 23 above)
Comment (adrianstepinski-cl) [reply server/server_test.go]: done

File: server/server.go [hunk 26]
```diff
@@ -400,6 +403,8 @@ func NewServer(configFile string) (*Server, error) {
 			SkipLogging: true,
 		}))
 	}
+	etcd_key := config.GetString("healthcheck/etcd_key", "/gohan")
```
[the remark this answers is not in the document]
Comment (adrianstepinski-cl) [reply server/server.go]: done

File: server/server_test.go (hunk 22 above)
Comment (adrianstepinski-cl) [reply server/server_test.go]: done

File: server/api.go [hunk 27]
```diff
@@ -210,6 +211,14 @@ func fillInContext(context middleware.Context, db db.DB,
 	context["openstack_client"] = identityService.GetClient()
 }
 
+func mustGetSchema(manager *schema.Manager, schemaID string) *schema.Schema {
+	schema, ok := manager.Schema(schemaID)
+	if !ok {
+		panic(fmt.Sprintf("The %s is missing. Check if gohan.json is loaded", schemaID))
```
[the remark this answers is not in the document]
Comment (adrianstepinski-cl) [reply server/api.go]: done

**Comment by p-kozlowski** [comment]
done in #716
